pkg/render/common/networkpolicy/networkpolicy.go

```diff
@@ -0,0 +1,201 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package networkpolicy
+
+import (
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v3 "github.com/tigera/api/pkg/apis/projectcalico/v3"
+	"github.com/tigera/api/pkg/lib/numorstring"
+)
+
+const TigeraComponentTierName = "allow-tigera"
+const TigeraComponentPolicyPrefix = TigeraComponentTierName + "."
+const TigeraComponentDefaultDenyPolicyName = TigeraComponentPolicyPrefix + "default-deny"
+
+var TCPProtocol = numorstring.ProtocolFromString(numorstring.ProtocolTCP)
+var UDPProtocol = numorstring.ProtocolFromString(numorstring.ProtocolUDP)
+var HighPrecedenceOrder = 1.0
+
+// AppendDNSEgressRules appends a rule to the provided slice that allows DNS egress. The appended rule utilizes label selectors and ports.
+func AppendDNSEgressRules(egressRules []v3.Rule, openShift bool) []v3.Rule {
+	if openShift {
+		egressRules = append(egressRules, []v3.Rule{
+			{
+				Action:   v3.Allow,
+				Protocol: &UDPProtocol,
+				Destination: v3.EntityRule{
+					NamespaceSelector: "projectcalico.org/name == 'openshift-dns'",
+					Selector:          "dns.operator.openshift.io/daemonset-dns == 'default'",
+					Ports:             Ports(5353),
+				},
+			},
+			{
+				Action:   v3.Allow,
+				Protocol: &TCPProtocol,
+				Destination: v3.EntityRule{
+					NamespaceSelector: "projectcalico.org/name == 'openshift-dns'",
+					Selector:          "dns.operator.openshift.io/daemonset-dns == 'default'",
+					Ports:             Ports(5353),
+				},
+			},
+		}...)
+	} else {
+		egressRules = append(egressRules, v3.Rule{
+			Action:   v3.Allow,
+			Protocol: &UDPProtocol,
+			Destination: v3.EntityRule{
+				NamespaceSelector: "projectcalico.org/name == 'kube-system'",
+				Selector:          "k8s-app == 'kube-dns'",
+				Ports:             Ports(53),
+			},
+		})
+	}
+
+	return egressRules
+}
+
+// CreateEntityRule creates an entity rule that matches traffic using label selectors based on namespace, deployment name, and port.
+func CreateEntityRule(namespace string, deploymentName string, port uint16) v3.EntityRule {
+	return v3.EntityRule{
+		NamespaceSelector: fmt.Sprintf("projectcalico.org/name == '%s'", namespace),
+		Selector:          fmt.Sprintf("k8s-app == '%s'", deploymentName),
+		Ports:             Ports(port),
+	}
+}
+
+// CreateSourceEntityRule creates a conventional entity rule that matches ingress traffic based on namespace and deployment name.
+func CreateSourceEntityRule(namespace string, deploymentName string) v3.EntityRule {
+	return v3.EntityRule{
+		Selector:          fmt.Sprintf("k8s-app == '%s'", deploymentName),
+		NamespaceSelector: fmt.Sprintf("name == '%s'", namespace),
```

Comment:
All of the existing ingress rules seem to use use `name` instead of `projectcalico.org/name`. Should this be changed?

Reply:
Is there a reason for our choice of "name" vs "projectcalico/name" to differ depending on whether the rule is ingress or egress? The pattern of the existing manifest appears to be "name" for ingress, and "projectcalico.org/name" for egress (hence the two different EntityRule-generating functions)

Reply:
I don't think that ingress vs egress makes a difference. This is probably just us being inconsistent. 

Being more consistent would be good - in general, it's better to use "real" labels than the fake labels we do behind the scenes because the real labels are viewable in the API.